app/uk/gov/hmrc/gform/graph/Recalculation.scala

```diff
@@ -166,9 +166,17 @@ class Recalculation[F[_]: Monad, E](
 
           noStateChange(evResult + (expr, exprResult))
       }
+
+      val validIfsResult: StateT[F, RecalculationState, Unit] =
+        evalValidIfs(evResult, recData, retrievals, booleanExprResolver, evaluationContext)
+
+      (validIfsResult, graphLayerResult).mapN { case (_, result) =>
+        result
+      }
```

Comment:
It is quite confusing what the purpose of this code is and why it works.

We should add comment to explain what is going on. ie. that we are only interested in `ValidIf` with `In` expression and that any other validIf is being ignored etc.

Reply:
Also perhaps we can simplify the code a bit too:

```
evalValidIfs(evResult, recData, retrievals, booleanExprResolver, evaluationContext).flatMap { _ =>
   graphLayerResult
}
```